### DISCUSSION TITLE
Can not import module of custom dll

### DISCUSSION THREAD

**Discussion by eragondr** [pythonnet/pythonnet#2385] [Q&A]
### Environment

-   Pythonnet version: 3.3.0
-   Python version:3.10
-   Operating System: win10
-   .NET Runtime: .Net 8

### Details

-   I was try to import my custom dll. At first I was import the full version of my C# dll. but it not works :  
- 
![image](https://github.com/pythonnet/pythonnet/assets/28474355/e718580e-7280-4709-b2ff-f5461bacec3c)

this is my full C# project 
![image](https://github.com/pythonnet/pythonnet/assets/28474355/dd02ec8c-39c6-4c69-a502-eb92c880893f)

then i remove all file and just create a demo class. 
![image](https://github.com/pythonnet/pythonnet/assets/28474355/b03f615f-18c8-4884-80a5-a6adbd922409)
i reimport and it works 
![image](https://github.com/pythonnet/pythonnet/assets/28474355/f4c40ab7-548c-474b-9cfe-ce910b86f4fc)

then i create a new folder call ResBaseArray and a Galois class
![image](https://github.com/pythonnet/pythonnet/assets/28474355/15bdc8c8-1a6c-489b-b70b-29bf9f23d448)

reimport 👍 
![image](https://github.com/pythonnet/pythonnet/assets/28474355/f6ee39b8-d9d6-48c5-b0af-8b38a9245c4a)
 `    from DesProduction import Demo
ModuleNotFoundError: No module named 'DesProduction'`
  
  How can i solve this please . thank for your works

**Comment by eragondr** [comment]
Any update? Please

**Comment by echidne** [comment]
Hi,
did you try with the full path?
`clr.AddReference(r'C:\PathToDllFolder\MyDll')`

**Reply by eragondr** [reply]
hi.
I did not try the full path but as you can see in the 4th picture i have put my dll inside the python packed and import. everything works  good.

**Reply by echidne** [reply]
Hi,
I tried to reproduce your problem without success.
What I did is to copy in a local "dll" directory a "working" library from my nugget directory. And use the same code than you without throwing a Error.
(What I mean by working was that if clr.AddReference("mynuggetdirectory/libraryname") then import(libraryname) all is working without throwing error.)
What happen if you do the same experience?
If thats works that means something is wrong with your library.
It it is the case the prob could perhaps come from the .net version you are targetting others are reported same throwed error
[here](https://stackoverflow.com/questions/68289729/pythonnet-cannot-access-to-dll-class), [here](https://stackoverflow.com/questions/65282510/is-net-core-or-net-5-0-supported-by-pythonnet), [here](https://stackoverflow.com/questions/67970591/pythonnet-dotnet-core-no-module-named), and https://github.com/pythonnet/pythonnet/issues/2148.
Myself I noticed that even with nugget installed library for .Net 5 I have the same kind of error than with your custom dll

**Reply by eragondr** [reply]
sorry for the late reply. this is how i do 

https://github.com/pythonnet/pythonnet/assets/28474355/3af8bfad-65eb-4501-af10-c065b9663d99

as you can see , everything was fine on the demo in the first time but when i copy my project source and re publish it give that error : No module named 'DesProduction'.

**Reply by echidne** [reply]
So If I understood correctly you first try with a basic dll worked but not when you copy your most complex code inside and publish again.
I confess I'm not a specialist of c# but one of the difference I noted is that in first basic dll your namespace have same name than your dll but not in the second try. Perhaps you could investigate in that way?
PS: I tried to reproduce your basic dll import but I was obliged to change the name of the namespace from DesProduction to DesProduction1 in Test.cs to have it working  as 
```
clr.AddReference("DesProduction")
from DesProduction1 import Test
```
so I suppose pythonnet import of dll is sensitive to name of namespace

**Reply by eragondr** [reply]
actually they both DesProduction. i was try with namespace "test" it still the same bro 
![image](https://github.com/pythonnet/pythonnet/assets/28474355/1902bd29-992b-4d4d-a224-ceeb22ef6e42)

**Reply by echidne** [reply]
If i read correctly its "DesProduction.RESBaseArray" not just "DesProduction" as in your test dll.
When you said you renamed as the namespace as "test" did you put in your python code?
```
clr.AddReference("DesProduction")
import test
```
for example if I create a test DesProduction.dll as you did but I change the namespacename in the test.cs:
![image](https://github.com/pythonnet/pythonnet/assets/65033881/e619d517-964e-46aa-a70e-3009ad39096d)

In my python code I have to write : 
```
import clr
clr.AddReference('DesProduction')
from DesProduction.RESBaseArray import Test
```
to have it working
if I write : 
```
from Desproduction import Test
```
I have as output : 
```
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[4], [line 1](vscode-notebook-cell:?execution_count=4&line=1)
----> [1](vscode-notebook-cell:?execution_count=4&line=1) from Desproduction import Test

ModuleNotFoundError: No module named 'Desproduction'
```
when you import a class you have to use the exact name of your namespace.

**Reply by eragondr** [reply]
thanks for your reply .
in my project i have many sub name space such as Desproduction.RESBaseArray  or  Desproduction.BaseArray or even Desproduction.RESBaseArray.ABC and so on. However the class i want to call in python is only in the name space Desproduction with no sub namespace such as class "Test" as you can see bro

**Reply by echidne** [reply]
> However the class i want to call in python is only in the name space Desproduction

So do you have a error message if you just import the namespace DesProduction as `import DesProduction`?

**Reply by eragondr** [reply]
> `So do you have a error message if you just import the namespace DesProduction as import DesProduction?`

yes i do bro. As you see in the video i kept the same code in python still using "Test" class, and in C# i have not change anything in "Test" class. all i do is copy the folder from my main source to C# project and retry with out any change of the code

**Reply by echidne** [reply]
Since the test code stop working after you copy Endcoder.cs and Decoder.cs in your Visual Studio project before publishing it, it's more likely that the problem is created by your csharp code. Did you try to introduce only one of them before publishing or part of them?

PS: probably unrelated but it exists a [numpy extension for Galois Field](https://github.com/mhostetter/galois).

**Reply by eragondr** [reply]
hmm . Maybe you're right . i will look into that. But is there any way i can debug it properly. Because my C# project was working fine on it own therefore i don't know how to debug it for pythonnet

**Comment by filmor** [comment]
Your C# code calls it `Endcoder`, in Python you have `Encoder`.

**Reply by eragondr** [reply]
thanks for your reply .
i do miss the typo but the thing is it sad ModleNotFoundError: No module named 'DesProduction' not Encoder

**Comment by DaveSkender** [comment]
If helpful, we have a similar setup and were able to successfully set this up.  You may need to define your path in a way that will survive the build.  See [our code](https://github.com/facioquo/stock-indicators-python/blob/a4a39ab515472bd64e7aa2eadcabb25131f9934b/stock_indicators/_cslib/__init__.py#L11-L26).
